python/cuml/test/test_random_forest.py

```diff
@@ -237,3 +235,97 @@ def test_rf_regression_default(datatype, column_info, nrows):
+    # Initialize, fit and predict using cuML's
+    # random forest classification model
+    cuml_model = curfr(seed=123, n_streams=1)
+    cuml_model.fit(X_train, y_train)
+
+    # predict using FIL
+    fil_preds_rerun = cuml_model.predict(X_test,
+                                         predict_model="GPU").copy_to_host()
+    cu_preds_rerun = cuml_model.predict(X_test, predict_model="CPU")
+    cu_r2_rerun = r2_score(y_test, cu_preds_rerun, convert_dtype=datatype)
+    fil_r2_rerun = r2_score(y_test, fil_preds_rerun, convert_dtype=datatype)
+    assert array_equal(fil_r2_orig, fil_r2_rerun, tol=1e-3)
+    assert array_equal(cu_r2_orig, cu_r2_rerun, tol=1e-3)
+    assert array_equal(fil_preds_orig, fil_preds_rerun, tol=1e-3)
+    assert array_equal(cu_preds_orig, cu_preds_rerun, tol=1e-3)
+
+
+@pytest.mark.parametrize('datatype', [np.float32])
+@pytest.mark.parametrize('column_info', [unit_param([16, 7]),
+                         quality_param([200, 100]),
+                         stress_param([500, 350])])
+@pytest.mark.parametrize('nrows', [unit_param(500), quality_param(5000),
+                         stress_param(500000)])
+def test_rf_classification_seed(datatype, column_info, nrows):
+
+    ncols, n_info = column_info
+    X, y = make_classification(n_samples=nrows, n_features=ncols,
+                               n_clusters_per_class=1, n_informative=n_info,
+                               random_state=0, n_classes=2)
+    X = X.astype(datatype)
+    y = y.astype(np.int32)
+    X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.8,
+                                                        random_state=0)
+
+    # Initialize, fit and predict using cuML's
+    # random forest classification model
+    cuml_model = curfc(seed=123, n_streams=1)
+    cuml_model.fit(X_train, y_train)
+
+    # predict using FIL
+    fil_preds_orig = cuml_model.predict(X_test,
+                                        predict_model="GPU").copy_to_host()
+    cu_preds_orig = cuml_model.predict(X_test,
+                                       predict_model="CPU")
+    cu_acc_orig = accuracy_score(y_test, cu_preds_orig)
+    fil_acc_orig = accuracy_score(y_test, fil_preds_orig)
+
+    # Initialize, fit and predict using cuML's
+    # random forest classification model
+    cuml_model = curfc(seed=123, n_streams=1)
+    cuml_model.fit(X_train, y_train)
+
+    # predict using FIL
+    fil_preds_rerun = cuml_model.predict(X_test,
+                                         predict_model="GPU").copy_to_host()
+    cu_preds_rerun = cuml_model.predict(X_test, predict_model="CPU")
+    cu_acc_rerun = accuracy_score(y_test, cu_preds_rerun)
+    fil_acc_rerun = accuracy_score(y_test, fil_preds_rerun)
+    assert array_equal(fil_acc_orig, fil_acc_rerun, tol=1e-3)
+    assert array_equal(cu_acc_orig, cu_acc_rerun, tol=1e-3)
+    assert array_equal(fil_preds_orig, fil_preds_rerun, tol=1e-3)
+    assert array_equal(cu_preds_orig, cu_preds_rerun, tol=1e-3)
```

Comment:
why test the reproducibility to 1e-3? Given that this is testing reproducibility of running the same algorithm/prediction with the same seed, the tolerance should be tighter (closer to float32 limitations at least), and the fact that there are sporadic failures in past CI logs, does this mean that the seeding might be buggy?

Reply:
`When i ran the code in my environment the condition was to check if they were equal and they would pass the tests but, when the CI failed i checked and it was due to difference in the predicted values. ` that is exactly my problem with this as it is, and as you mention it might be a bug in the seeding. It might have been luck it worked in your environment and not in CI, setting a seed should make things reproducible as much as possible within the same environment. 

`They would have slightly different decimal values which would cause the scores to change` sure, but the tolerance value of 1e-3 for that seems way too loose for a seed repeatability test, even with single precision. Maybe it is not, but that is what I would like to know, why the choice of 1e-3? Given that this style of seed test will be spread to more algos over time, I think we need a good baseline for a minimum standard of precision (which might not be the same for all algos as well), which needs some thought.

Reply:
While running the tests the r2 scores and the predicted values would fluctuate 3 to 4 decimal places and so i set the tol to 1e-3

Reply:
Sure, but my question is whether the fact that fluctuating 3 to 4 places is the acceptable behavior or is it just showing lack of reproducibility? @JohnZed maybe you have some input as well about that?